enhancements/machine-config/install-time-support-for-image-mode-on-openshift.md

```diff
@@ -0,0 +1,343 @@
+
+### Goals
+
+- Enable cluster admins to use Image Mode on OpenShift at install time
+- User creates a customized OS container image to be used when a cluster is installed
+- Maintain OpenShift's single-click upgrade experience maintaining customizations across upgrades
+- Ensure all Machine Config Pools (control plane, worker, any custom pools) can all use custom OS images
+- Provide a declarative approach to OS customization through MachineOSConfig CRD
+
+### Non-Goals
+
+- Replacing the existing ignition-based configuration system entirely
+- Supporting arbitrary OS distributions beyond RHCOS-based images
+- Providing external CI/CD integration for image builds
+- Enabling the custom OS container image to be built during bootstrap
+
+## Proposal
+
+### Workflow Description
+
+The install time Image Mode on OpenShift workflow integrates with the existing OpenShift installation process in the following steps.
+
+#### Step 1: Installation Configuration with Pre-Built Container Image
+
+1. Cluster admin builds the customized OS container image and pushes it to a registry to be used at install time
+2. Cluster admin adds the `pull secret` needed to pull the container image to the cluster's `global pull-secret`
+3. Cluster admin runs `openshift-install create manifests` to generate the base Kubernetes manifests
+4. Cluster admin places `MachineOSConfig` (MOSC) YAML files in the `<install-dir>/manifests/` directory for any MachineConfigPools they want to use the Image Mode workflow for
+    - MOSC has annotation pointing to the pre-built container image (seeding the image to be used at install time)
+    - MOSC has the custom Containerfile used when building the custom OS container image in step 1
+    - MOSC has `pushSecret` set with the correct credentials needed to push any new images built post install to the registry
+5. Cluster admin runs `openshift-install create cluster` which automatically:
+   - Creates ignition configs by base64 encoding all YAML files from `manifests/` directory
+
+#### Step 2: Bootstrap Phase
+
+1. Bootstrap MCO recognizes MachineOSConfig manifests and stores them for later API creation
+2. Bootstrap MCO creates rendered MachineConfigs for each pool
+3. Bootstrap MCO then does the following to inject the pre-built container image into the rendered MC
+    - Maps MachineOSConfigs with pre-built annotations to their target pools
+    - Validates pre-built image format (digest format required)
+    - Sets `OSImageURL = preBuiltImage` in rendered MachineConfigs for pools with pre-built images and MachineOSConfigs
+    - Adds tracking annotation to modified MachineConfigs
+4. Bootstrap MCO creates ignition files with modified rendered MCs containing `OSImageURL` pointing to the pre-built container image
+5. Bootstrap MCO writes MachineOSConfig manifests to `/etc/mcs/bootstrap/machine-os-configs/` for post-bootstrap API creation
```

Comment:
Is this step necessary? I don't think we explicitly write anything to disk for API creation in the MCS generally speaking

Reply:
@yuqi-zhang How does the bridging between bootstrap and the final cluster happen for e.g. day-1 MachineConfigs? I would assume that same mechanism should work for MOSCs too.

Reply:
So if you're asking about render config generation, those are rendered separately, but I assume you're asking if I provide a MachineConfig during install time via the manifest folder, how that gets processed?

In the bootstrap node that entire directory gets copied in and generated by the bootstrap controller to serve in the MCS. Then (bootkube?) applies that manifest as well into the "real cluster" api when it comes up. The MCO doesn't directly process any of those files I believe.

Reply:
This is true, we don't need to explicitly write the MachineOSConfig to disk. I have updated the steps accordingly